### PR TITLE
feat: rate limit with rln configuration

### PR DESCRIPTION
**Pull request by kaichaosun**
# Description

Support RLN contract's configuration like: allow 100 messages per 1 minute.

~**[DO NOT MERGE] Use this PR to test 1:1 with RLN**~
The default rate limiter is used when not configured, so this PR is safe to be merged.

# Changes

- [x] create a new custom rate limiter with refill at certain interval strategy

### CODE REVIEW

**Comment by chaitanyaprem** [comment]
@kaichaosun i am guessing this is for some sort of testing and analysis because RLN already the logic to accept/reject msgs based on rate limit, if so better not to merge it. Would it be required for both lightclient and relay?

Would be helpful if you can give more context as to what this change is meant to be used for.
 

Also iiuc RLN integration is supposed to be done for 1:1 chats first right. if so, wondering if we should introduce this limiter at such lower layer or a higher layer in status-go where we can distinguish 1:1 chats and community chats.

File: waku/v2/api/publish/message_sender.go [hunk 1]
```diff
@@ -89,7 +86,7 @@ func NewMessageSender(publishMethod PublishMethod, publisher Publisher, logger *
 	return &MessageSender{
 		publishMethod: publishMethod,
 		publisher:     publisher,
-		rateLimiter:   NewPublishRateLimiter(DefaultPublishingLimiterRate, DefaultPublishingLimitBurst),
+		rateLimiter:   NewRlnRateLimiter(RlnLimiterCapacity, RlnLimiterRefillInterval),
```
[the remark this answers is not in the document]
Comment (chaitanyaprem) [reply waku/v2/api/publish/message_sender.go:89]: maybe we can use some build flag to choose between default or RLN Rate limiter?

**Review by chaitanyaprem — COMMENTED** [review]
Rate limiter logic looks good to me, but i am surprised golang's default rate-limiter doesn't support any other interval other than rate/sec.

**Comment by richard-ramos** [comment]
Interesting implementation of a rate limiter!

**Review by plopezlpz — APPROVED** [review]

File: waku/v2/api/publish/rln_rate_limiting.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	}
+
+	return false
+}
+
+func (rl *RlnRateLimiter) Check(ctx context.Context, logger *zap.Logger) error {
+	if rl.Allow() {
+		return nil
+	}
+	logger.Error("could not send message rate limited", zap.Error(ErrRateLimited))
```
Comment (richard-ramos) [inline waku/v2/api/publish/rln_rate_limiting.go:62]: Consider not logging this as there's already a log in https://github.com/status-im/status-go/blob/develop/wakuv2/message_publishing.go#L104 for when message fails to publish.
You would end up with:
```
2024-12-05T21:11:22.668Z    ERROR  somefile.go:1547  could not send message rate limited {"error": "rate limit exceeded"}
2024-12-05T21:11:22.668Z    ERROR  somefile.go:1547  could not send message {"envelopeHash": "0xSomeHash", "pubsubTopic": "someTopic", "contentTopic": "someContentTopic", "timestamp": 1234567890, "error": "rate limit exceeded"}
```

**Review by richard-ramos — APPROVED** [review]

**Comment by plopezlpz** [comment]
relates to https://github.com/status-im/status-go/issues/6130

**Pull request merged**